### PR TITLE
[Exec] Improve documentation 

### PR DESCRIPTION
**Pull request by Josar**
Added information how to test if user openhab is able to execute the command.
Added generalized full setup example.
Change Minimal example, also made all items to Strings as the exec binding only returns strings.

This is an attempt to overcome the remarks in #2795

Signed-off-by: Josua Arndt josuaarndt@live.de (github: josar)

### CODE REVIEW

**Comment by martinvw** [comment]
@Josar can you please rebase on the master?

**Comment by Josar** [comment]
@martinvw hope rebase worked.

File: addons/binding/org.openhab.binding.exec/README.md [hunk 1]
```diff
@@ -60,8 +69,79 @@ Thing exec:command:myscript [command="php ./configurations/scripts/script.php %2
...
+      
+      // If the returned string is just a number it can be parsed
+      // If not a regex or another transformation can be used
+      YourNumber.postUpdate( 
+            (Integer::parseInt(yourcommand_out.state.toString) as Number ) 
+      ) 
+end
+
+```
+The logging massages can be viewed in the Karaf console have a closer look [in the manual](http://docs.openhab.org/administration/console.html) for more information
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.exec/README.md]: Typo massages

File: addons/binding/org.openhab.binding.exec/README.md [hunk 2]
```diff
@@ -60,8 +69,79 @@ Thing exec:command:myscript [command="php ./configurations/scripts/script.php %2
...
+
+## Source
+[OpenHAB 1 Addons wiki](https://github.com/openhab/openhab1-addons/wiki/Raspberry-Pi-System-Temperature)
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.exec/README.md]: Please add a newline before and after each heading in a markdown file, otherwise some of the implementations including the one used by [docs.openhab.org](https://docs.openhab.org) might fail to render properly.

File: addons/binding/org.openhab.binding.exec/README.md [hunk 3]
```diff
@@ -29,16 +34,18 @@ For each command a separate Thing has to be defined.
 Thing exec:command:apc [command="/usr/local/bin/apcaccess  status", interval=15, timeout=5, autorun=false]
 ```
 
-`command` itself can be enhanced using the well known syntax of the **java.util.Formatter** class.
+```command``` itself can be enhanced using the well known syntax of the **java.util.Formatter** class.
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.exec/README.md]: Why did you change this, imho single backticks are fine for inline 'code'

File: addons/binding/org.openhab.binding.exec/README.md (hunk 2 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.exec/README.md]: I doubt whether we want to keep pointing to the old wiki, @ThomDietrich wdyt?

**Comment by martinvw** [comment]
I think it's okay, @ThomDietrich can you also quickly look at it :-)

File: addons/binding/org.openhab.binding.exec/README.md [hunk 4]
```diff
@@ -10,13 +10,18 @@ Currently, the binding supports a single type of Thing, being the ```command```
 
 The binding does not require any specific configuration.
 
-Note that the commands are executed in the context and with the privileges of the process running the java virtual machine.
-It is not advised to run the virtual machine as superuser/root.
+
+Note that the commands are executed in the context and with the privileges of the process running the java virtual machine. It is not advised to run the virtual machine as superuser/root. Linux Os needs the user openhab/openhabian to be able to execute dedicated command, it is advised to always test in the command line if this is possible.
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: I believe it's more common to capitalize  Java Virtual Machine. Also I wonder if you'd like to add a few words of explanation like: "in which openHAB is executed."

I'd make the new sentence a new paragraph and write:

```text
**Linux:**
On a Linux system the system user `openhab` needs to have the privileges needed to execute your intended  command.
It is advised to test the correct operation of the command in the scope of the `openhab` user by on the command line:

File: addons/binding/org.openhab.binding.exec/README.md [hunk 5]
```diff
@@ -10,13 +10,18 @@ Currently, the binding supports a single type of Thing, being the ```command```
...
+```
+sudo -u openhab <YOUR COMMAND>
+```
 
 ## Thing Configuration
 
-The `command` Thing requires the command to execute on the shell.
-Optionally one can specify:
+
+The ```command``` Thing requires the command to execute on the shell. Optionally one can specify:
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: As @martinvw said, these extra backticks are not needed. In fact I'd write "command" Thing rather than `command` Thing.

Please do not merge sentences but keep them on separate lines. Markdown will merge them in the produced html pages automatically.

File: addons/binding/org.openhab.binding.exec/README.md [hunk 6]
```diff
@@ -29,16 +34,18 @@ For each command a separate Thing has to be defined.
 Thing exec:command:apc [command="/usr/local/bin/apcaccess  status", interval=15, timeout=5, autorun=false]
 ```
 
-`command` itself can be enhanced using the well known syntax of the **java.util.Formatter** class.
+`command` itself can be enhanced using the well known syntax of the **java.util.Formatter** class.
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: Here `command` is the right way to write it, other than for the "command" Thing

Please remove the space at the end.

File: addons/binding/org.openhab.binding.exec/README.md [hunk 7]
```diff
@@ -29,16 +34,18 @@ For each command a separate Thing has to be defined.
 Thing exec:command:apc [command="/usr/local/bin/apcaccess  status", interval=15, timeout=5, autorun=false]
 ```
 
-`command` itself can be enhanced using the well known syntax of the **java.util.Formatter** class.
+`command` itself can be enhanced using the well known syntax of the **java.util.Formatter** class. 
 The following parameters are automatically added:
 
+
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: Please remove every second empty line

File: addons/binding/org.openhab.binding.exec/README.md [hunk 8]
```diff
@@ -60,8 +67,81 @@ Thing exec:command:myscript [command="php ./configurations/scripts/script.php %2
 **demo.items**
 
 ```
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: Please write:

    ```java

File: addons/binding/org.openhab.binding.exec/README.md [hunk 9]
```diff
@@ -60,8 +67,81 @@ Thing exec:command:myscript [command="php ./configurations/scripts/script.php %2
...
+
+String APCRaw "[%s]" (All) {channel="exec:command:apc:output"} 
+String APCRunning { channel="exec:command:apc:run"}
+String APCExitValue {channel="exec:command:apc:exit"}
+String APCLastExecution {channel="exec:command:apc:lastexecution"}
+```
+
+## Full Example
+
+Following is an example how to set up an exec Thing, debug it with a rule and set the returned string to an Number Item.
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: exec command Thing

File: addons/binding/org.openhab.binding.exec/README.md [hunk 10]
```diff
@@ -60,8 +67,81 @@ Thing exec:command:myscript [command="php ./configurations/scripts/script.php %2
...
+String APCRunning { channel="exec:command:apc:run"}
+String APCExitValue {channel="exec:command:apc:exit"}
+String APCLastExecution {channel="exec:command:apc:lastexecution"}
+```
+
+## Full Example
+
+Following is an example how to set up an exec Thing, debug it with a rule and set the returned string to an Number Item. 
+
+**For this to work also the openhab-transformation-regex has to be installed**
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: the openHAB RegEx Transformation

File: addons/binding/org.openhab.binding.exec/README.md [hunk 11]
```diff
@@ -60,8 +67,81 @@ Thing exec:command:myscript [command="php ./configurations/scripts/script.php %2
...
+{
+        Frame {
+            Switch item=YourTrigger
+            Text item=YourNumber
+        }
+}
 ```
+
+**demo.rules**
+```
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: Add "java" syntax highlighting to all of these code fences.

File: addons/binding/org.openhab.binding.exec/README.md [hunk 12]
```diff
@@ -60,8 +67,81 @@ Thing exec:command:myscript [command="php ./configurations/scripts/script.php %2
...
+
+## Full Example
+
+Following is an example how to set up an exec Thing, debug it with a rule and set the returned string to an Number Item. 
+
+**For this to work also the openhab-transformation-regex has to be installed**
+
+**demo.things**
+```
+Thing exec:command:yourcommand [ command="<YOUR COMMAND> %2$s"
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: ` %2$s`?

File: addons/binding/org.openhab.binding.exec/README.md [hunk 13]
```diff
@@ -60,8 +67,81 @@ Thing exec:command:myscript [command="php ./configurations/scripts/script.php %2
...
+
+Following is an example how to set up an exec Thing, debug it with a rule and set the returned string to an Number Item. 
+
+**For this to work also the openhab-transformation-regex has to be installed**
+
+**demo.things**
+```
+Thing exec:command:yourcommand [ command="<YOUR COMMAND> %2$s"         
+                                 interval=0,
+                                 autorun=true ]
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: Please condense to one line

File: addons/binding/org.openhab.binding.exec/README.md [hunk 14]
```diff
@@ -60,8 +67,81 @@ Thing exec:command:myscript [command="php ./configurations/scripts/script.php %2
...
+      
+      // If the returned string is just a number it can be parsed
+      // If not a regex or another transformation can be used
+      YourNumber.postUpdate( 
+            (Integer::parseInt(yourcommand_out.state.toString) as Number ) 
+      ) 
+end
+
+```
+The logging messages can be viewed in the Karaf console have a closer look [in the manual](http://docs.openhab.org/administration/console.html) for more information
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: Please remove this sentence. It is common knowledge, incomplete and not specific to the binding

File: addons/binding/org.openhab.binding.exec/README.md (hunk 12 above)
Comment (Josar) [reply addons/binding/org.openhab.binding.exec/README.md]: state of the input channel

File: addons/binding/org.openhab.binding.exec/README.md (hunk 10 above)
Comment (Josar) [reply addons/binding/org.openhab.binding.exec/README.md]: But it is called like this in Karaf when i try to install it!

Comment (ThomDietrich) [reply addons/binding/org.openhab.binding.exec/README.md]: Yes but that's just the technical package name. In Paper UI or in documentation it has a human readable name. In the last sentence I used the name "Paper UI" not its feature name `openhab-ui-paper`, that would be unintuitive.

The Console (karaf) is just one tool to manage openHAB and documentation should stay general and end user oriented. The end user needn't know the feature name, if he/she wants, it's easily found out via `feature:list` in said console.

File: addons/binding/org.openhab.binding.exec/README.md [hunk 15]
```diff
@@ -4,64 +4,148 @@ This binding integrates the possibility to execute arbitrary shell commands.
...
+
+```shell
+sudo -u openhab <YOUR COMMAND>
+```
 It is not advised to run the virtual machine as superuser/root.
 
 ## Thing Configuration
 
-The `command` Thing requires the command to execute on the shell.
+The "command" thing requires the command to execute on the shell.
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: Thing is a proper noun and should always be uppercase

File: addons/binding/org.openhab.binding.exec/README.md [hunk 16]
```diff
@@ -4,64 +4,148 @@ This binding integrates the possibility to execute arbitrary shell commands.
...
 For each command a separate Thing has to be defined.
 
-```
-Thing exec:command:apc [command="/usr/local/bin/apcaccess  status", interval=15, timeout=5, autorun=false]
+```java
+Thing exec:command:uniquename [command="/command/to/execute here", interval=15, timeout=5, autorun=false]
 ```
 
-`command` itself can be enhanced using the well known syntax of the **java.util.Formatter** class.
+The `command` itself can be enhanced using the well known syntax of the **java.util.Formatter** class.
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: Even better, let's do what we did over in https://docs.openhab.org/configuration/items.html#state-presentation and turn the formatter info into a link

Comment (ThomDietrich) [reply addons/binding/org.openhab.binding.exec/README.md]:
```
Formatting of the presentation is done applying [Java formatter class syntax](http://docs.oracle.com/javase/7/docs/api/java/util/Formatter.html#syntax).
```

File: addons/binding/org.openhab.binding.exec/README.md [hunk 17]
```diff
@@ -4,64 +4,148 @@ This binding integrates the possibility to execute arbitrary shell commands.
...
 
-The `command` Thing requires the command to execute on the shell.
+The "command" thing requires the command to execute on the shell.
 Optionally one can specify:
 
--   a transformation to apply on the execution result,
--   an interval, in seconds, the command will be repeatedly executed,
--   a time-out, in seconds, the execution of the command will time out, and lastly,
--   a boolean parameter to make the command execute immediately every time the state of the input channel has changed.
+- `transform` - A transformation to apply on the execution result,
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: Please link transformation to: https://docs.openhab.org/addons/transformations.html

File: addons/binding/org.openhab.binding.exec/README.md [hunk 18]
```diff
@@ -4,64 +4,148 @@ This binding integrates the possibility to execute arbitrary shell commands.
...
 ```
 
-`command` itself can be enhanced using the well known syntax of the **java.util.Formatter** class.
+The `command` itself can be enhanced using the well known syntax of the **java.util.Formatter** class. 
 The following parameters are automatically added:
 
 -   the current date (as java.util.Date, example: `%1$tY-%1$tm-%1$td`)
 -   the current State of the input channel (see below, example: `%2$s`)
 
+**For this to work also the openhab-transformation-regex has to be installed**
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: This sentence is now in the article twice

File: addons/binding/org.openhab.binding.exec/README.md [hunk 19]
```diff
@@ -4,64 +4,148 @@ This binding integrates the possibility to execute arbitrary shell commands.
...
+```java
+Thing exec:command:uniquename [command="/command/to/execute here", interval=15, timeout=5, autorun=false]
 ```
 
-`command` itself can be enhanced using the well known syntax of the **java.util.Formatter** class.
+The `command` itself can be enhanced using the well known syntax of the **java.util.Formatter** class. 
 The following parameters are automatically added:
 
 -   the current date (as java.util.Date, example: `%1$tY-%1$tm-%1$td`)
 -   the current State of the input channel (see below, example: `%2$s`)
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: The example doesn't become clear. Why the second argument?

File: addons/binding/org.openhab.binding.exec/README.md (hunk 18 above)
Comment (Josar) [reply addons/binding/org.openhab.binding.exec/README.md]: I don't thinnk they are automatticaly added. The transformation service has to be installed for them to work. Or they wont work. I don't know where to put it best. But i am sure it should be here. And i am sure if it is not stated at the example this also leads to confusion.

As nobody who starts working with the Binding will connect the dots from there to the example.

File: addons/binding/org.openhab.binding.exec/README.md [hunk 20]
```diff
@@ -39,29 +47,108 @@ The following parameters are automatically added:
...
+Following is an example how to set up an exec command thing, debug it with a rule and set the returned string to an Number Item. 
+
+**For this to work also the openHAB RegEx Transformation has to be installed**
+
+**demo.things**
+
+```java
+// "%2$s" will be replace by the input channel, this makes it possible to use one command line with different arguments.
+// e.g: "ls" as <YOUR COMMAND> and "-a" or "-l" as additional argument set to the input channel in the rule.
+Thing exec:command:yourcommand [ command="<YOUR COMMAND> %2$s", interval=0, autorun=false ]
```
Comment (ThomDietrich) [inline addons/binding/org.openhab.binding.exec/README.md]: Are you sure it is explicitly needed to be added? I did not do any testing but I'd think the input channel is automatically added to the end of the command during execution!?

Comment (Josar) [reply addons/binding/org.openhab.binding.exec/README.md]: yes it is. I use it all the time like this. And executing the command without it just executes the command! Without adding the input channel.
https://community.openhab.org/t/openhab2-exec-binding/27819/5?u=josar

**Comment by martinvw** [comment]
@Josar can you rebase and take a look at the conflicts. Then we have a great addition, thanks for your help and not giving up!

File: addons/binding/org.openhab.binding.exec/README.md (hunk 20 above)
Comment (ThomDietrich) [reply addons/binding/org.openhab.binding.exec/README.md]: Okay then, thanks!

**Comment by ThomDietrich** [comment]
Same goes for me. Thanks for keeping up with my requests. You've reached a nice addition!
